pkg/datapath/probes/probes.go

```diff
@@ -0,0 +1,355 @@
+				return err
+			}
+			if _, err = stderrBuf.WriteTo(infoFile); err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+	return
+}
+
+// writeFeatures wrtites definitions of features into the header file which is
+// going to be used by BPF programs.
+func writeFeatures(featureFile, infoFile, warningFile io.Writer, probesDir, libIncludeDir, outDir string) error {
+	if _, err := io.WriteString(featureFile, `#ifndef BPF_FEATURES_H_
+#define BPF_FEATURES_H_
+`); err != nil {
+		return err
+	}
+	if err := probeKernelConfig(infoFile, warningFile); err != nil {
+		return fmt.Errorf("kernel config probe failed: %s", err)
+	}
+	if err := probeRunLl(featureFile, infoFile, probesDir, libIncludeDir, outDir); err != nil {
+		return fmt.Errorf("ll probe failed: %s", err)
+	}
+	if _, err := io.WriteString(featureFile, "#endif /* BPF_FEATURES_H_ */\n"); err != nil {
+		return err
+	}
+	return nil
+}
+
+// RunProbes runs probes for BPF features. It generates a header file for BPF
+// programs and log files.
+func RunProbes(bpfDir, stateDir string) error {
+	var featureFile, infoFile, warningFile *fileWithCleanup
+	var outDir string
+	var errs *multierror.Multierror
+	var err error
+
+	probesDir := path.Join(bpfDir, "probes")
+	libIncludeDir := path.Join(bpfDir, "include")
+	featureFilePath := path.Join(stateDir, "globals", "bpf_features.h")
+	infoFilePath := path.Join(stateDir, "bpf_features.log")
+	warningFilePath := path.Join(stateDir, "bpf_requirements.log")
+
+	os.Remove(infoFilePath)
+	os.Remove(warningFilePath)
+
+	featureFile, err = openOrCreate(featureFilePath)
+	if err != nil {
+		errs = multierror.Append(errs, err)
+		goto cleanup
+	}
+	infoFile, err = openOrCreate(infoFilePath)
+	if err != nil {
+		errs = multierror.Append(errs, err)
+		goto cleanup
+	}
+	warningFile, err = openOrCreate(warningFilePath)
+	if err != nil {
+		errs = multierror.Append(errs, err)
+		goto cleanup
+	}
+
+	outDir, err = ioutil.TempDir("", "cilium-probes-out")
+	if err != nil {
+		err = fmt.Errorf("could not create temporary directory for probes: %s", err)
+		errs = multierror.Append(errs, err)
+		goto cleanup
+	}
+	defer os.RemoveAll(outDir)
+
+	if err := writeFeatures(featureFile, infoFile, warningFile, probesDir, libIncludeDir, outDir); err != nil {
+		errs = multierror.Append(errs, err)
+		goto cleanup
+	}
+
+	if err := featureFile.Sync(); err != nil {
+		errs = multierror.Append(errs, err)
+		goto cleanup
+	}
+	if err := infoFile.Sync(); err != nil {
+		errs = multierror.Append(errs, err)
+		goto cleanup
+	}
+	if err := warningFile.Sync(); err != nil {
+		errs = multierror.Append(errs, err)
+		goto cleanup
+	}
+
+cleanup:
+	if err := featureFile.Close(); err != nil {
+		errs = multierror.Append(errs, err)
+	}
+	if err := infoFile.Close(); err != nil {
+		errs = multierror.Append(errs, err)
+	}
+	if err := warningFile.Close(); err != nil {
+		errs = multierror.Append(errs, err)
+	}
+
+	if errs != nil {
+		return errs
```

Comment:
I think the multierror thing here is just to allow any failures during the `Close()` operations to be logged, but it kinda complicates the return because now we need this multierror thing. However, the caller is unlikely to be able to do anything specific about a failure to `Close()`. In fact, all it does is log the message and ignore it. Wouldn't it be simpler to just log any `Close()` failures here, and propagate back the original error?

Reply:
Done, I dropped multierror in favor of logging all errors and returning the last one.